primal-sieve/src/streaming/primes.rs

```diff
@@ -80,18 +77,49 @@ impl Primes {
         Primes {
             early: Early::Two,
             base: 0,
-            current: iter.next().unwrap(),
-            elems: iter,
-            streaming: streaming,
+            ones,
+            streaming,
             sieving_primes: sieving,
             left_over: None,
         }
     }
+
+    #[inline]
+    fn from_bit_index(&self, i: usize) -> Option<usize> {
+        let w = wheel::from_bit_index(i);
+        self.base.checked_add(w)
+    }
+
+    fn advance_ones(&mut self) -> bool {
+        let low = self.streaming.low;
+        let high = low.saturating_add(streaming::SEG_LEN);
+
+        for q in self.left_over.into_iter().chain(self.sieving_primes.as_mut().unwrap()) {
+            if q.saturating_mul(q) >= high {
+                self.left_over = Some(q);
+                break
+            }
+            if q > streaming::isqrt(streaming::SEG_LEN) {
+                self.streaming.add_sieving_prime(q, low);
+            }
+        }
+
+        match self.streaming.next() {
+            Some((_, bits)) => {
+                self.base = low;
+                self.ones = bits.clone().into_ones();
+                true
+            },
+            None => false,
+        }
+    }
 }
 
+
 impl Iterator for Primes {
```

Comment:
Given this implementation and `SievePrimes` is almost identical, I'm curious on your thoughts about merging them?

Reply:
Hmm. The code _is_ almost identical, but they are structurally different enough that they can't just be composed. Maybe an internal helper trait is due here.

Reply:
Maybe worth if/as more methods are specialised, but could be done as eventual follow up work to keep the scope of this smaller (if you're wanting to get it polished off).

Reply:
I added cross-reference comments as a reminder, but otherwise I think it's OK for now.